lib/sms/sms.c

```diff
@@ -0,0 +1,322 @@
+
+/** @brief Check if the received AT notification is a +CMT event. */
+static bool sms_is_cmd_notification(const char *const at_notif)
+{
+	if (at_notif == NULL) {
+		return false;
+	}
+
+	if (strlen(at_notif) <= AT_SMS_NOTIFICATION_LEN) {
+		return false;
+	}
+
+	return (strncmp(at_notif, AT_SMS_NOTIFICATION,
+			AT_SMS_NOTIFICATION_LEN) == 0);
+}
+
+/** @brief Parse the +CMT unsolicited received message in PDU mode. */
+static int sms_cmt_notif_parse(const char *const buf)
+{
+	/* Parse the received message. */
+	int err = at_parser_max_params_from_str(buf, NULL, &resp_list,
+						AT_CMT_PARAMS_COUNT);
+
+	if (err != 0) {
+		LOG_ERR("Unable to parse CMT notification, err=%d", err);
+		return err;
+	}
+
+	/* Check the AT response format. */
+	if (at_params_valid_count_get(&resp_list) != AT_CMT_PARAMS_COUNT) {
+		LOG_ERR("Invalid CMT notification format");
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+/** @brief Save the SMS notification parameters. */
+static int sms_cmt_notif_save(void)
+{
+	if (cmt_rsp.alpha != NULL) {
+		k_free(cmt_rsp.alpha);
+	}
+
+	if (cmt_rsp.pdu != NULL) {
+		k_free(cmt_rsp.pdu);
+	}
+
+	/* Save alpha as a null-terminated String. */
+	size_t alpha_len;
+	(void)at_params_size_get(&resp_list, 1, &alpha_len);
+
+	cmt_rsp.alpha = k_malloc(alpha_len + 1);
+	if (cmt_rsp.alpha == NULL) {
+		return -ENOMEM;
+	}
+	(void)at_params_string_get(&resp_list, 1, cmt_rsp.alpha, &alpha_len);
+	cmt_rsp.alpha[alpha_len] = '\0';
+
+	/* Length field saved as number. */
+	(void)at_params_short_get(&resp_list, 2, &cmt_rsp.length);
+
+	/* Save PDU as a null-terminated String. */
+	size_t pdu_len;
+	(void)at_params_size_get(&resp_list, 3, &pdu_len);
+	cmt_rsp.pdu = k_malloc(pdu_len + 1);
+	if (cmt_rsp.pdu == NULL) {
+		return -ENOMEM;
+	}
+
+	(void)at_params_string_get(&resp_list, 3, cmt_rsp.pdu, &pdu_len);
+	cmt_rsp.pdu[pdu_len] = '\0';
+
+	return 0;
+}
+
+/** @brief Handler for AT responses and unsolicited events. */
+void sms_at_handler(void *context, char *at_notif)
+{
+	ARG_UNUSED(context);
+	LOG_DBG("AT notification '%s'", log_strdup(at_notif));
+
+	/* Check if the notification is a CMT event. */
+	if (!sms_is_cmd_notification(at_notif)) {
+		LOG_DBG("Ignore notifications other than CMT");
+		return;
+	}
+
+	/* Parse and validate the CMT notification, then extract parameters. */
+	if (sms_cmt_notif_parse(at_notif) != 0) {
+		LOG_ERR("Invalid CMT notification");
+		return;
+	}
+
+	/* Extract and save the SMS notification parameters. */
+	int valid_notif = sms_cmt_notif_save();
+
+	/* Ack the SMS PDU in any case. */
+	int ret = at_cmd_write(AT_SMS_PDU_ACK, NULL, 0, NULL);
+
+	if (ret != 0) {
+		LOG_ERR("Unable to ACK the SMS PDU");
+	}
+
+	if (valid_notif != 0) {
+		LOG_ERR("Invalid SMS notification format");
+		return;
+	}
+
+	/* Notify all subsribers. */
+	LOG_DBG("Valid SMS notification decoded");
+	for (sms_handle_t i = 0; i < ARRAY_SIZE(subsribers); i++) {
+		const struct sms_subscriber *s = subsribers + i;
```

Comment:
subsribers[i]

and change
sms_handle_t i
to
int i
or
uint32_t i

I wouldn't bother making a variable for subsribers[i], so just get rid of the variable s and inline subsribers[i].

Reply:
The compiler will inline it for me, but I will change it.